docs/src/main/asciidoc/native-reference.adoc

```diff
@@ -0,0 +1,1382 @@
+
+The time spent in analysis and compilation phases depends on how big the application is.
+The bigger the application, the longer it takes to compile it.
+However, there are certain features that can have an exponential effect.
+For example, when registering types and methods for reflection access,
+the analysis can’t easily see what’s behind those types or methods,
+so it has to do more work to complete the analysis step.
+
+=== Why is runtime performance of natives slower compared to JVM mode?
+
+As with most things in life there are some trade offs involved when choosing native compilation over JVM mode.
+So depending on the application the runtime performance of a native application might be slower compared to JVM mode,
+though that’s not always the case.
+
+JVM execution of an application includes runtime optimization of the code that profits from profile information built up during execution.
+That includes the opportunities to inline a lot more of the code,
+locate hot code on direct paths (i.e. ensure better instruction cache locality)
+and cut out a lot of the code on cold paths (on the JVM a lot of code does not get compiled until something tries to execute it -- it is replaced with a trap that causes deoptimization and recompilation).
+Removal of cold paths provides many more optimization opportunities than are available for ahead of time compilation because it significantly reduces the branch complexity and combinatorial logic of the smaller amount of hot code that is compiled.
+
+By contrast, native executable compilation has to cater for all possible execution paths when it compiles code offline since it does not know which are the hot or cold paths and cannot use the trick of planting a trap and recompiling if it is hit. For the same reason it cannot load the dice to ensure that code cache conflicts are minimized by co-locating hot paths adjacent.
+Native executable generation is able to remove some code because of the closed world hypothesis but that is often not enough to make up for all the benefits that profiling and runtime deopt & recompile provides to the JVM JIT compiler.
+
+Note, however, that there is a price you pay for that potentially higher JVM speed, and that price is in increased resource usage (both CPU and memory) and startup time because:
+
+1. it takes some time before the JIT kicks in and fully optimizes the code.
+2. the JIT compiler consumes resources that could be utilized by the application.
+3. the JVM has to retain a lot more metadata and compiler/profiler data to support the better optimizations that it can offer.
+
+The reason for 1) is that code needs to be run interpreted for some time and, possibly, to be compiled several times before all potential optimizations are realized to ensure
+a) that it’s worth compiling that code path, i.e. it’s being executed enough times, and
+b) that we have enough profiling data to perform meaningful optimizations.
```

Comment:
You aware that all of this will be on the same line once interpreted by Asciidoctor?

Reply:
Will reformat accordingly.